src/msg_recorder/recorder.h

```diff
@@ -49,7 +91,16 @@ void MessageRecorder::RegisterChannel(const MessageRecorder::channel_subid_t sub
 
     this->Subscribe<message_t>(
         subid,
-        [record_file](const std::shared_ptr<message_t>& message) { record_file->Write(MessageToStr(*message)); },
+        [this, record_file, alias](const std::shared_ptr<message_t>& message) {
+            if (!data_copied_) {
+                std::unique_lock lock(recorder_mtx_);
```

Comment:
Try [`cris::core::JobRunnerStrand`](https://github.com/cyfitech/cris-core/blob/857125196b374b88d1e5860616c43251dfd7c9a8/src/sched/job_runner.h#L14L15).
It serialize the jobs so that you don't need to worry about concurrency of the stranded jobs.

`allow_concurrency` is a simplified use case of strands. It only serialize the incoming messages. If you want more strict serialization, you may need to do manually. See unit tests if you need examples.